### PR TITLE
Improve `object_store crate` documentation

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-rs/issues/2202

# Rationale for this change
 
More docs `==` more better

# What changes are included in this PR?

- [x] Update main crate readme (What appears in https://crates.io/crates/object_store)
- [x] Add main example (front page of https://docs.rs/object_store/latest/object_store/)

Example front page:

Screen Shot 2022-08-01 at 8 36 08 AM

# Are there any user-facing changes?

Better docs

### CODE REVIEW

File: object_store/README.md [hunk 1]
```diff
@@ -19,8 +19,21 @@
...
-Originally developed for [InfluxDB IOx](https://github.com/influxdata/influxdb_iox/) and later split out and donated to Apache Arrow.
+Using this crate, the same binary and code can easily run in multiple
+clouds and local test environments, via a simple runtime configuration
+change. Supported object stores include:
+
+* [AWS S3](https://aws.amazon.com/s3/)
+* [Azure Blob Storage](https://azure.microsoft.com/en-us/services/storage/blobs/)
+* [Google Cloud Storage](https://cloud.google.com/storage)
+* Local files
+* Memory,
```
Comment (tustvold) [inline object_store/README.md:33]:
```suggestion
* Memory
```

File: object_store/README.md [hunk 2]
```diff
@@ -19,8 +19,21 @@
 
 # Rust Object Store
 
-A crate providing a generic interface to object stores, such as S3, Azure Blob Storage and Google Cloud Storage.
+A focused, easy to use, idiomatic, high performance, `async` object
+store library interacting with object stores.
```
Comment (tustvold) [inline object_store/README.md:23]:
```suggestion
A high performance, `async` object store crate that provides a
uniform interface for interacting with various kinds of object store
```

I think all the qualifiers is a bit much :laughing:

File: object_store/src/lib.rs [hunk 3]
```diff
@@ -28,15 +28,128 @@
...
+//!
+//! # async fn example() {
+//! use std::sync::Arc;
+//! use object_store::{path::Path, ObjectStore};
+//! use futures::stream::StreamExt;
+//!
+//! // create an ObjectStore
+//! let object_store: Arc<dyn ObjectStore> = Arc::new(get_object_store());
+//!
+//! // List all files in the 'data' path.
```
Comment (tustvold) [inline object_store/src/lib.rs:71]:
```suggestion
//! // List all files below the 'data' path.
```

File: object_store/src/lib.rs [hunk 4]
```diff
@@ -28,15 +28,128 @@
...
+//!      })
+//!      .await;
+//! # }
+//! ```
+//!
+//! Which will print out something like the following:
+//!
+//! ```text
+//! Name: data/file01.parquet, size: 112832
+//! Name: data/file02.parquet, size: 143119
```
Comment (tustvold) [inline object_store/src/lib.rs:99]:
```suggestion
//! Name: data/file01.parquet, size: 112832
//! Name: data/file02.parquet, size: 143119
//! Name: data/child/file03.parquet, size: 100
```

I think it is worth highlighting that this is "recursive"

**Review by tustvold — APPROVED** [review]
Minor nits, thank you for doing this

File: object_store/README.md (hunk 2 above)
Comment (alamb) [reply object_store/README.md:23]: I am practicing "marketing" -- I would like to run an experiment to see how much attention I can garner with a buzzword laden lead compared to a more utilitarian, less self-aggrandizing, summary.

File: object_store/src/lib.rs (hunk 4 above)
Comment (alamb) [reply object_store/src/lib.rs:99]: great point

**Review by Ted-Jiang — APPROVED** [review]
Nice write up! 😊

**Comment by alamb** [comment]
Thanks @Ted-Jiang  !

**Pull request merged**